examples/two_moons_classification_sngp.pct.py

```diff
@@ -0,0 +1,228 @@
+
+class SNGPDeepFeatureExtractorSubNet(WithSpectralNorm, DeepResidualFeatureExtractorSubNet):
+    pass
+
+# %% [markdown]
+# Then, we can define our SNGP model by:
+#
+# - Changing the deep feature extractor: from `MLPDeepFeatureExtractorSubNet` to `SNGPDeepFeatureExtractorSubNet`
+# - Changing the output layer: from `Dense` to `RandomFeatureGaussianProcess`.
+
+# %%
+from fortuna.model.utils.random_features import RandomFeatureGaussianProcess
+
+from fortuna.model.sngp import SNGPMixin
+class SNGPModel(SNGPMixin, DeepResidualNet):
+    def setup(self):
+        if len(self.widths) != len(self.activations):
+            raise Exception(
+                "`widths` and `activations` must have the same number of elements."
+            )
+        self.dfe_subnet = SNGPDeepFeatureExtractorSubNet(
+            dense=self.dense,
+            widths=self.widths,
+            activations=self.activations[:-1],
+            dropout=self.dropout,
+            dropout_rate=self.dropout_rate,
+            spectral_norm_bound=self.spectral_norm_bound,
+            spectral_norm_iteration=self.spectral_norm_iteration,
+        )
+        self.output_subnet = RandomFeatureGaussianProcess(
+            features=self.output_dim,
+            hidden_features=self.gp_hidden_features,
+            normalize_input=self.normalize_input,
+            covmat_kwargs={
+            "ridge_penalty": self.ridge_penalty,
+            "momentum": self.momentum,
+        },
+        )
+
+    def __call__(self, x, train=False, **kwargs):
+        x = self.dfe_subnet(x, train)
+        x = self.output_subnet(x, return_full_covmat=self.use_full_covmat)
+        return x
+
+# %%
+model = SNGPModel(
+        output_dim=output_dim,
+        activations=(nn.relu, nn.relu, nn.relu, nn.relu, nn.relu, nn.relu),
+        widths=(128,128,128,128,128,128),
+        dropout_rate=0.1,
+        gp_hidden_features=1024,
+        spectral_norm_bound=0.9,
+        ridge_penalty=1,
+    )
+
+# %%
+from functools import partial
+import jax.numpy as jnp
+
+from fortuna.model.model_manager.classification import SNGPModelManager
+from fortuna.prob_model.callbacks.sngp import ResetCovarianceCallback
+from fortuna.prob_model.prior import IsotropicGaussianPrior
+
+prob_model = ProbClassifier(
+    model=model,
+    prior=IsotropicGaussianPrior(log_var=jnp.log(1./1e-4) - jnp.log(TRAIN_DATA_SIZE)),
+    posterior_approximator=MAPPosteriorApproximator(),
+    output_calibrator=None,
+    model_manager_cls=partial(SNGPModelManager, mean_field_factor=1),
+)
+status = prob_model.train(
+    train_data_loader=train_data_loader,
+    val_data_loader=val_data_loader,
+    calib_data_loader=val_data_loader,
+    fit_config=FitConfig(
+        monitor=FitMonitor(metrics=(accuracy,)),
+        optimizer=FitOptimizer(method=optax.adam(1e-4), n_epochs=100),
+        callbacks=[ResetCovarianceCallback(
+            precision_matrix_key_name='precision_matrix',
+            ridge_penalty=1
+        )]
+    )
+)
+
```

Comment:
I would echo Gianluca's comment re. the background knowledge required for some of the arguments. Could we perhaps add some discussion on there purpose in the notebook so that the reader can learn about their purpose, if desired.

Reply:
I changed the API. Now a user can train an SNGP model by:

- Passing a deep feature extractor (with Spectral Norm) directly to the `ProbModel`'s `train` method.
- Selecting `SNGPPosteriorApproximator` as the `posterior_approximator` for the `ProbModel`'s `train` method. `SNGPPosteriorApproximator` has one single required argument (`output_dim`), all the other args are optional and set to a default.